### PR TITLE
Fix advertise/publish race condition

### PR DESCRIPTION
**Pull request by jtbandes**
**User-Facing Changes**
Fixed an issue where extension panels were unable to advertise and immediately publish on the same channel.

**Description**
Addresses some of the issues discussed in https://github.com/foxglove/studio/issues/5455 & https://github.com/joshnewans/fox-pub-error-mwe

The MessagePipeline keeps a list of `publishersById` which holds the publishers for each panel. It also merges them into a single list of publishers, and calls `player.setPublishers()` with the merged list. The merging was being done in a reducer, and the `setPublishers` call was happening in a `useEffect`. However, the `useEffect` could occur "too late" and a panel (such as https://github.com/joshnewans/fox-pub-error-mwe) could call `publish()` before the advertisement had actually been set on the player, which would throw an error.

To address the timing issue, I moved `setPublishers` out of a `useEffect` and it now happens synchronously during the `advertise` call. Since the list of publishers was not used outside of the MessagePipeline for any other purpose, I've also moved it out of the "`public`" state.

The Rosbridge player would also throw an error if a publish() occurred before the advertisements had been fully set up (after a connection was established). However, the panels are not aware of the state of the underlying websocket connection, so they have no way to avoid publishing before it's fully established. So I added logic to avoid throwing an error in this case. I wasn't able to reproduce the same issue with the websocket player so I didn't make any change to it.

### CODE REVIEW

File: packages/studio-base/src/components/MessagePipeline/index.test.tsx [hunk 1]
```diff
@@ -484,30 +482,48 @@ describe("MessagePipelineProvider/useMessagePipeline", () => {
...
-    const lastPublishers = result.current.publishers;
+    const lastPublishers = player.publishers;
     // cause the player to emit a frame outside the render loop to trigger another render
     await doubleAct(async () => await player.emit());
     // make sure publishers are reference equal when they don't change
-    expect(result.current.publishers).toBe(lastPublishers);
+    expect(player.publishers).toBe(lastPublishers);
+  });
+
+  it("keeps publishers when changing player", () => {
```
Comment (snosenzo) [inline packages/studio-base/src/components/MessagePipeline/index.test.tsx:509]: a comment to explain why this is tested against would be useful, not super evident from title.

**Review by snosenzo — COMMENTED** [review]
not very familiar with the zustand change implications, so might want to get some other eyes on that

File: packages/studio-base/src/components/MessagePipeline/index.test.tsx (hunk 1 above)
Comment (jtbandes) [reply packages/studio-base/src/components/MessagePipeline/index.test.tsx]: updated the title, hopefully it is more clear now

**Review by snosenzo — APPROVED** [review]

File: packages/studio-base/src/components/MessagePipeline/store.ts [hunk 2]
```diff
@@ -93,14 +95,14 @@ export function createMessagePipelineStore({
...
       messageEventsBySubscriberId: new Map(),
       subscriptions: [],
-      publishers: [],
       sortedTopics: [],
       datatypes: new Map(),
       setSubscriptions(id, payloads) {
         get().dispatch({ type: "update-subscriber", id, payloads });
       },
       setPublishers(id, payloads) {
         get().dispatch({ type: "set-publishers", id, payloads });
```
Comment (defunctzombie) [inline packages/studio-base/src/components/MessagePipeline/store.ts:104]: What's the purpose of this dispatch now that the publishers are set internally?

File: packages/studio-base/src/components/MessagePipeline/store.ts [hunk 3]
```diff
@@ -93,14 +95,14 @@ export function createMessagePipelineStore({
...
+        get().player?.setPublishers(get().allPublishers);
```
Comment (defunctzombie) [inline packages/studio-base/src/components/MessagePipeline/store.ts:105]: How is `allPublishers` updated prior to calling `setPublishers` on the player?

File: packages/studio-base/src/components/MessagePipeline/store.ts (hunk 2 above)
Comment (jtbandes) [reply packages/studio-base/src/components/MessagePipeline/store.ts:104]: We still need to keep track of the state somewhere.

File: packages/studio-base/src/components/MessagePipeline/store.ts (hunk 3 above)
Comment (jtbandes) [reply packages/studio-base/src/components/MessagePipeline/store.ts:105]: It's updated as part of the reducer call from the `dispatch` above

**Comment by jtbandes** [comment]
> what's with the // foxglove-depcheck-used: babel-plugin-transform-import-meta ? was this causing an issue previously?

I added `babel-plugin-transform-import-meta` to the studio-base deps to ease development in the private repo. That for some reason triggered the unused dependencies linter where previously it was not triggered.

File: packages/studio-base/src/components/MessagePipeline/store.ts (hunk 3 above)
Comment (defunctzombie) [reply packages/studio-base/src/components/MessagePipeline/store.ts:105]: I did some reading on this and I think dispatch returns the new state? It wasn't entirely clear to me. But if it does that might be a little clearer here. Something to be aware of for the future.

**Pull request merged**